qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,286 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""SteppableOptimizer interface"""
+
+from abc import abstractmethod, ABC
+from dataclasses import dataclass
+from typing import Union, Callable, Optional, Tuple, List
+from .optimizer import Optimizer, POINT, OptimizerResult
+
+CALLBACK = Callable
+
+
+@dataclass
+class AskObject(ABC):
+    """Base class for return type of :meth:`~.SteppableOptimizer.ask`.
+
+    Args:
+        x_fun: Point or list of points where the function needs to be evaluated to compute the next
+        state of the optimizer.
+        x_jac: Point or list of points where the gradient/jacobian needs to be evaluated to compute the
+        next state of the optimizer.
+
+    """
+
+    x_fun: Optional[Union[POINT, List[POINT]]] = None
+    x_jac: Optional[Union[POINT, List[POINT]]] = None
+
+
+@dataclass
+class TellObject(ABC):
+    """Base class for argument type of :meth:`~.SteppableOptimizer.tell`.
+
+    Args:
+        eval_fun: Image of the function at :attr:`~.Ask_Object.x_fun`.
+        eval_jac: Image of the gradient-jacobian at :attr:`~.Ask_Object.x_jac`.
+
+    """
+
+    eval_fun: Union[float, List[float], None] = None
+    eval_jac: Union[POINT, List[POINT], None] = None
+
+
+@dataclass
+class OptimizerState:
+    """Base class representing the state of the optimizer.
+
+    Any variable that changes during the optimization should be stored in this dataclass.
```

Comment:
What does this really mean? This class has the set defined below so presumably an optimizer would at least be expected to save as state here any of these - i.e. maintain its state using an instance of this

Reply:
I am not really sure what do you mean by this comment. What I meant to say is that this class will store any variable that will change during the optimization process.
For example, ``tol`` is a float that is used to stop the optimization process once the objective function has a value below it. Therefore this variable doesn't change from one step to the next and is, therefore, an attribute of the optimizer.
The ``stepsize`` however, is something that evolves with every step and forms part of the ``state`` of the optimizer.

Reply:
So what you mean perhaps is that if the optimizer has additional state variables beyond the set "in this dataclass" (using the phrase at the end) the developer should create a sub-class of this with the additional variables so that its state can be properly managed?

Reply:
The information stored inside this dataclass contains everything related to a specific optimization process. Imagine we have a gradient descent instance.
This instance might have some tolerance, some perturbation and some function that yields the learning rate. This is always going to be the same for every minimization that we run.
Now, with that gradient_descent instance we might choose to minimize several different functions, or the same function with different initial points and so on.
Everything stored inside of OptimizerState will be different for each one of this cases, but the "hyper-parameters" are part of the gradient_descent instance and will not change at any time.
Maybe with an example it would be easier.
``` python
gradient_descent = GradientDescent(maxiter = 100, tol = 0.01, perturbation = 0.1) # This are hyper-parameters

gradient_descent.start(fun=f1,x0=init_point_1)
for _ in range(20):
    gradient_descent.step()
state1 = gradient_descent.state

gradient_descent.start(fun=f1,x0=init_point_2)
for _ in range(20):
    gradient_descent.step()
state2 = gradient_descent.state

gradient_descent.start(fun=f2,x0=init_point_3)
for _ in range(20):
    gradient_descent.step()
state3 = gradient_descent.state

#Now we choose to finish the optimization on state2
gradient_descent.state = state2
while gradient_descent.continue_condition():
    gradient_descent.step()
result=gradient_descent.create_result()
```
Now, at the end of this script we have 2 different states that we can choose to keep optimizing if we want. Inside of the state there is the relevant information in order to perform a next step. The rest of the parameters that we gave to gradient_descent are just describing how to perform the next step given a state.

Reply:
Would this be more precise? 🙂 
```suggestion
    This class stores the current state of the optimizer, given by the current point and
    (optionally) information like the function value, the gradient or the number of 
    function evaluations. This dataclass can also store any other individual variables that
    change during the optimization.
```

Reply:
Changed!